### PR TITLE
Add kubernetes-plugins ci initial test

### PR DESCRIPTION
**Pull request by ryanohnemus**
Adds an initial CI test folder with a single test to check if the kubernetes filter plugin is enhancing records with k8s pod labels. 

We can add kubernetes_namespace label checks after https://github.com/fluent/fluent-bit/pull/8279 has merged.

NOTE: Since the fluentbit container does not have tar available, I wasn't able to just use a file output with a `kubectl cp` and instead just use another output plugin (elasticsearch) to be able to fetch and check the results easily. (Obviously this will look very similar to the elasticsearch-basic ci/bats test).

### CODE REVIEW

**Comment by patrick-stephens** [comment]
> NOTE: Since the fluentbit container does not have tar available, I wasn't able to just use a file output with a `kubectl cp` and instead just use another output plugin (elasticsearch) to be able to fetch and check the results easily. (Obviously this will look very similar to the elasticsearch-basic ci/bats test).

Yeah I think that's fair enough, I was thinking it potentially could be done with stdout and then checking the output of the pods or another output like Loki, etc. (we could also use a PV for file access) but this is fine and works.

Just need to check why it did not run the integration tests in this case. We may need to run them for the specific branch but I was hoping it would trigger on PR to confirm the test actually passes before we merge.

**Comment by ryanohnemus** [comment]
@patrick-stephens I like your idea of using stdout plugin much better than my original version. I just pushed a new update and removed the elasticsearch output dependency. 

This version works by creating fluentbit that will read container logs that are in the `$TEST_NAMESPACE`. Since I'm running fluentbit as a deployment, any test container I create needs to be on the same k8s node for fluentbit to see it's logs. I use the fluent-bit container's nodeName after it's deployment to specify that as a nodeSelector when I run a hello world app (just a base alpine image that prints `hello world` every 2 seconds). That hello world app is being assigned a custom label `this_is_a_test_label=true` and after all that is deployed and running, I look for it specifically.

Let me know your thoughts!

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:14]: Let's do this on every test I think so call it `setup` and `teardown` rather than once-per-file - I think that was for the ES tests as it is relatively expensive to run up ES.

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:28]: As above, rename to `teardown` for future updates to add more tests and do it per-test

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:58]: we should `assert_success` after this otherwise failures are ignored

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:61]: Remove the `run` at the start so it'll fail if the command does, otherwise we need to add an `assert_success` afterwards

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:65]: I'm not a fan of sleeps and either way we should make sure the pod stays running as well afterwards - a simple `kubectl get pod -l ..` just to verify it exists

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:70]: Should this always succeed? If so no need for the `run` prefix although you're using the output so maybe an `assert_success` afterwards.

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:76]: Messy, I'd be inclined to use an `assert_output --partial` just to handle all the corner cases already handled by the assertion library but that will fail then whereas we want to retry so I think this is the best we can do.

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:81]: let's add more detail to the output on the selector and string we're after - saves people diving into the test code to figure it out

File: tests/kubernetes-plugins/resources/fluentbit-basic.yaml
Comment (patrick-stephens) [inline tests/kubernetes-plugins/resources/fluentbit-basic.yaml:53]: I think we want to remove the ES output

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:61]: Let's run this up first to speed up deployment so it's there ready to go when the FB deployment comes up.

**Review by patrick-stephens — COMMENTED** [review]
Some minor niggles but otherwise looks good to me.

File: tests/kubernetes-plugins/basic.bats
Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:65]: removed the sleep, using the `try` in the same way we're using it to check for fluent-bit itself

File: tests/kubernetes-plugins/basic.bats
Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:61]: I originally had issues with using a second `try` check... it appears we need to update our bash-detik version because multiple `try` statements to find different pods in the same test were not working. Updating that library fixes the issue, so I have left this run as is, then added: 

```
    try "at most 15 times every 5s " \
        "to find 1 pods named 'hello-world-1' " \
        "with 'status' being 'Running'"
```
which is the same thing we are doing for fluent-bit container check.

Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:61]: this depends on the fluent-bit container's nodeName, so we need to run it after.

File: tests/kubernetes-plugins/basic.bats
Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:81]: when I clean this up and just use a `assert_output --partial` it'll be easier to see.

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [reply tests/kubernetes-plugins/basic.bats:61]: Ah my bad, can you make sure to add a comment to explain that then for the invariable "optimisation" someone does later :)

Comment (patrick-stephens) [reply tests/kubernetes-plugins/basic.bats:61]: ah right, we can update - it's been stuck for a while on this version but there's no inherent reason to pin it

Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:62]: it's already there above the `run` command on line 57 :), haha

**Comment by ryanohnemus** [comment]
@patrick-stephens - ok i just pushed the latest version. 

bats-detik version was updated, it looked like some of the other libs had higher versions but i only updated the one i needed to fix that `try` bug. I believe I addressed your other comments or responded to them, please take another look, hopefully it's looking much cleaner now.

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [reply tests/kubernetes-plugins/basic.bats:62]: can we update this or we need the `assert_success` afterwards

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:60]: sanity check for zero length output or any other weirdness that kubectl get could return without failing?

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:75]: remove the `run` or add `assert_success` afterwards

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:80]: we do have a wait for output function actually we could reuse from our Calyptia one rather than having to rely on sleep always working: https://github.com/calyptia/bats/blob/main/test-helpers.bash#L121

We can probably copy over the test helpers to here to use it

Comment (patrick-stephens) [reply tests/kubernetes-plugins/basic.bats:80]: ah, ignore me - that's for local containers and not pods

Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:80]: That looks like it's just for the container's log itself, we can get that pretty quickly (ie seeing the output of the fluent-bit startup/version text). This is a mildly different as the sleep is specifically to ensure that the input `tail` plugin has had enough time for the configured `Refresh_Interval` so that it can pick up the container log for the new hello-world-1 container and process it. We could look into a future improvement that adds a k8s service for the http service in fluent and then use a curl/etc to get the metrics that tells you the input plugin has processed records, but the `sleep` for now doesn't feel that bad to me 😄

Comment (patrick-stephens) [reply tests/kubernetes-plugins/basic.bats:83]: let's make the sleep plenty long then, say 10s just to ensure we definitely have it and prevent any flakes at the expense of test time

File: tests/kubernetes-plugins/basic.bats
Comment (patrick-stephens) [inline tests/kubernetes-plugins/basic.bats:16]: only possible issue I can see here is with the occasional namespace delete blocked on finalizers... 

However there's no good solution for that.

File: tests/kubernetes-plugins/basic.bats
Comment (ryanohnemus) [reply tests/kubernetes-plugins/basic.bats:85]: done! just force pushed the last update

**Review by patrick-stephens — APPROVED** [review]
I think this is fine, I'd like to verify it runs successfully though  - I'm not sure why the PR is not running it but we can potentially explicitly run the workflow for a branch or if you can confirm it is ok manually?

**Comment by patrick-stephens** [comment]
@ryanohnemus does this pass locally for you btw?

**Comment by ryanohnemus** [comment]
@patrick-stephens 

```
./run-tests.sh tests/kubernetes-plugins/basic.bats   
+++ dirname /Users/ryan/repos/fluent-bit-ci/tools/install-bats.sh
++ cd /Users/ryan/repos/fluent-bit-ci/tools
++ pwd
+ SCRIPT_DIR=/Users/ryan/repos/fluent-bit-ci/tools
+ BATS_ROOT=/Users/ryan/repos/fluent-bit-ci/tools/bats
+ BATS_FILE_ROOT=/Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-file
+ BATS_SUPPORT_ROOT=/Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-support
+ BATS_ASSERT_ROOT=/Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-assert
+ BATS_DETIK_ROOT=/Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-detik
+ rm -rf /Users/ryan/repos/fluent-bit-ci/tools/bats
+ mkdir -p /Users/ryan/repos/fluent-bit-ci/tools/bats/lib
+ BATS_ASSERT_VERSION=2.0.0
+ BATS_SUPPORT_VERSION=0.3.0
+ BATS_FILE_VERSION=0.3.0
+ BATS_DETIK_VERSION=1.2.1
++ mktemp -d
+ DOWNLOAD_TEMP_DIR=/var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR
+ pushd /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR
/var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR ~/repos/fluent-bit-ci
+ curl -LO https://github.com/bats-core/bats-assert/archive/refs/tags/v2.0.0.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 21631    0 21631    0     0  42141      0 --:--:-- --:--:-- --:--:-- 42141
+ unzip -q v2.0.0.zip
+ mv -f /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR/bats-assert-2.0.0 /Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-assert
+ rm -f v2.0.0.zip
+ curl -LO https://github.com/bats-core/bats-support/archive/refs/tags/v0.3.0.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 18035  100 18035    0     0  27533      0 --:--:-- --:--:-- --:--:-- 27533
+ unzip -q v0.3.0.zip
+ mv -f /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR/bats-support-0.3.0 /Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-support
+ rm -f v0.3.0.zip
+ curl -LO https://github.com/bats-core/bats-file/archive/refs/tags/v0.3.0.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 47242    0 47242    0     0  73563      0 --:--:-- --:--:-- --:--:--  474k
+ unzip -q v0.3.0.zip
+ mv -f /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR/bats-file-0.3.0 /Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-file
+ rm -f v0.3.0.zip
+ curl -LO https://github.com/bats-core/bats-detik/archive/refs/tags/v1.2.1.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  154k    0  154k    0     0   304k      0 --:--:-- --:--:-- --:--:--  304k
+ unzip -q v1.2.1.zip
+ mv -f /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR/bats-detik-1.2.1/lib /Users/ryan/repos/fluent-bit-ci/tools/bats/lib/bats-detik
+ rm -f v1.2.1.zip
+ popd
~/repos/fluent-bit-ci
+ rm -rf /var/folders/m3/_vmlrw313xddy2ynj_g9b7680000gn/T/tmp.A5D8LenhmR


========================
Starting tests.
========================

Fluentbit repository: ghcr.io/fluent/fluent-bit - tag: latest


1..1
ok 1 test fluent-bit reads k8s labels in 118000ms


========================
All tests passed!
========================

```

**Comment by patrick-stephens** [comment]
Can you rebase and push a new commit to your original PR so it can trigger the new tests?

**Comment by ryanohnemus** [comment]
@patrick-stephens - ok i created a new draft pr https://github.com/fluent/fluent-bit-ci/pull/122 over here, but i still don't see it automatically kicking off a run?

**Comment by patrick-stephens** [comment]
Sorry I meant to your Fluent Bit PR - the namespace label support there.

**Pull request merged**
